File: src/main/java/io/scriptor/conf/ArrayNode.java
package io.scriptor.conf;

import com.carrotsearch.hppc.ObjectArrayList;
import com.carrotsearch.hppc.cursors.ObjectCursor;
import org.jetbrains.annotations.NotNull;

import java.util.Iterator;

public final class ArrayNode implements Node<Object>, Iterable<ObjectCursor<Node<?>>> {

    private final ObjectArrayList<Node<?>> list = new ObjectArrayList<>();

    public ArrayNode() {
    }

    public @NotNull ArrayNode add(final @NotNull Node<?> value) {
        list.add(value);
        return this;
    }

    public @NotNull ArrayNode remove(final int index) {
        list.removeAt(index);
        return this;
    }

    @Override
    public <N extends Node<?>> @NotNull N get(final @NotNull Class<N> type, final int index) {
        return type.cast(list.get(index));
    }

    @Override
    public @NotNull Iterator<ObjectCursor<Node<?>>> iterator() {
        return list.iterator();
    }
}

File: src/main/java/io/scriptor/conf/ArrayNode.kt
package io.scriptor.conf

import kotlin.reflect.KClass
import kotlin.reflect.cast

class ArrayNode : Node<Any>, Iterable<Node<*>> {

    private val list: MutableList<Node<*>> = ArrayList()

    fun add(value: Node<*>) {
        list.add(value)
    }

    fun remove(index: Int) {
        list.removeAt(index)
    }

    override operator fun <N : Node<*>> get(type: KClass<N>, index: Int): N {
        return type.cast(list[index])
    }

    override operator fun iterator(): MutableIterator<Node<*>> {
        return list.iterator()
    }
}
